### ISSUE TITLE
hdf5::property::CloseDegree::STRONG does not work as expeded

### ISSUE THREAD

**Issue by jkotan** [ess-dmsc/h5cpp#346]
the issue was split from #345. It looks like setting CloseDegree does not work. The following test

    TEST_F(FileCreation, test_closeattr_withoutclose_strong)
    {
      hdf5::property::FileAccessList fapl;
      // fapl.driver(file::PosixDriver());
      fapl.close_degree(hdf5::property::CloseDegree::STRONG);
      hdf5::property::FileCreationList fcpl;
      EXPECT_EQ(fapl.close_degree(), hdf5::property::CloseDegree::STRONG);
      auto nexus_file = file::create("testclose.h5",
				     file::AccessFlags::TRUNCATE, fcpl, fapl);
      std::string hdf5_version =  "1.0.0";
      auto type = datatype::create<std::string>();
      dataspace::Scalar space;

      auto r1 = nexus_file.root();
      r1.attributes.create("HDF5_version", type, space).write(hdf5_version);
      r1.close();

      EXPECT_EQ(nexus_file.count_open_objects(file::SearchFlags::ALL), 1);
      nexus_file.close();

      hdf5::property::FileAccessList fapl2;
      fapl2.close_degree(hdf5::property::CloseDegree::STRONG);
      EXPECT_EQ(fapl2.close_degree(), hdf5::property::CloseDegree::STRONG);
      auto file2 = hdf5::file::open("testclose.h5",
				    hdf5::file::AccessFlags::READONLY, fapl2);
      auto r2 = file2.root();

      auto attr = r2.attributes[0];
      // attr.close(); with this works
      EXPECT_EQ(file2.count_open_objects(file::SearchFlags::ALL), 3);

      r2.close();

      EXPECT_NO_THROW(hdf5::file::open("testclose.h5",
				       hdf5::file::AccessFlags::READWRITE));
    }

fails on the last opening file.

**Comment by jkotan** [comment]
No. I would rather say it is a problem in h5cpp implementation of links.
Even if is a bug of hdf5 the h5cpp library should also work with older and CURRENT versions of hdf libs.
Beauty of the code cannot be more important than functionality. Otherwise no one ones to use it.
BTW,  at DESY this feature is already hardly used (I've added as a patch to debian package on request)

**Comment by eugenwintersberger** [comment]
Have you tried the feature with the plain C-API?

**Comment by eugenwintersberger** [comment]
Just had a look on the link implementation: we hold a reference to the parent object of the attribute. But as the link contains only a path and the file where the parent object is stored I do not see how this would influence the closing behavior. I could imagine that this causes issues if we would hold an ID to the parent but if that would be the case than what would be the case of the STRONG closing policy?

**Comment by jkotan** [comment]
The problem is in file objects of links. If you look at the hdf5 lib code you see that it assumes that when you open a file you use only one copy of file identifier object, i.e. "STRONG closing" is applied only to one instance, takes care only about its children. (I think it may have to do with MPI implementation). So I suppose it is a feature.
With the plain C-API there is no problem (I tested it) because usually you create only one file object.
Therefore I've added a loop which decreases the file identifier counter on strong closing.

A much more elegant solution would be: instead of file objects in links keep only filename objects in links, i.e. simple string without file identifier (it is not needed).
But it requires rewriting a part of the code.

**Comment by eugenwintersberger** [comment]
Ok. I got the point.The problem is the open file handler in the Link class. We need a way to work around this.
